Feat/met 3128 record parsing and augmenting

jochen-vermeulen opened the pull request.

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/AbstractReferenceTerm.java:
```diff
@@ -1,40 +1,40 @@
     }
-
-    if(other == null || getClass() != other.getClass()){
+    if (!(otherObject instanceof ReferenceTerm)) {
```
Why change this?
It's comparing here the interface type which it shouldn't and it's not symmetrical.
I think sonar will complain for this too

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/AbstractSearchTerm.java:
```diff
@@ -1,46 +1,45 @@
     }
-
-    if(other == null || getClass() != other.getClass()){
+    if (!(otherObject instanceof SearchTerm)) {
```
Same here

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/SearchTerm.java:
```diff
@@ -1,14 +1,35 @@
 
   String getLanguage();
 
+  /**
```
Equals and hashCode shouldn't be in the interface imo.
The implementer should decide if the equals should be applied on the implementing class or not.
Btw they were removed on a previous review.

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/ReferenceTerm.java:
```diff
@@ -2,11 +2,32 @@
+  Set<EntityType> getCandidateTypes();
 
   URL getReference();
+
```
Same here 

"Equals and hashCode shouldn't be in the interface imo.
The implementer should decide if the equals should be applied on the implementing class or not.
Btw they were removed on a previous review."

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/ReferenceTermContext.java:
```diff
@@ -17,22 +20,11 @@ public ReferenceTermContext(URL reference, Set<FieldType> fieldTypes) {
+    return fieldTypes.stream().map(FieldType::getEntityType).collect(Collectors.toSet());
   }
 
-  @Override
```
I think equals and hashcode should be declared here too, preferably with using the new fields or just by using super

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/SearchTermContext.java:
```diff
@@ -17,27 +19,11 @@ public SearchTermContext(String textValue, String language, Set<FieldType> field
+    return Collections.unmodifiableSet(fieldTypes);
   }
-
-  @Override
```
Same here 

"I think equals and hashcode should be declared here too, preferably with using the new fields or just by using super"

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/SearchTermImpl.java:
```diff
@@ -0,0 +1,23 @@
+  @Override
+  public Set<EntityType> getCandidateTypes() {
+    return Collections.unmodifiableSet(entityTypes);
+  }
```
Same here

"I think equals and hashcode should be declared here too, preferably with using the new fields or just by using super"

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/ReferenceTermImpl.java:
```diff
@@ -0,0 +1,24 @@
+  @Override
+  public Set<EntityType> getCandidateTypes() {
+    return Collections.unmodifiableSet(entityTypes);
+  }
```
Same here:

"I think equals and hashcode should be declared here too, preferably with using the new fields or just by using super"

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/AbstractReferenceTerm.java:
```diff
@@ -1,40 +1,40 @@
   @Override
-  public boolean equals(Object other){
-    if(other == this){
+  public final boolean equals(Object otherObject) {
```
Why final here, were are enforcing equals/hashcode in the hierarchy which I believe we should not

stzanakis commented on metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/internal/AbstractSearchTerm.java:
```diff
@@ -1,46 +1,45 @@
   @Override
-  public boolean equals(Object other){
-    if(other == this){
+  public final boolean equals(Object otherObject) {
```
Same here:

"Why final here, were are enforcing equals/hashcode in the hierarchy which I believe we should not"

stzanakis reviewed (changes requested):
My comments mostly reflect code principles, more specifically for equals/hashcode where I think we should not transparently enforce implementations of it. But rather give freedom to the implementer. The fact that we use/implement them in a specific manner shouldn't need to be enforced to another implementation.
Any strong opinions on this?

jochen-vermeulen commented:
The implementation of equals and hashcode were made to comply with an old version of the EntityResolver interface (where it was important for the input objects to strictly satisfy equality conditions). But I changed the interface later to use generics, so that these strict conditions are no longer needed (but they were kept in place).

I now removed them, in this case that's fine. But perhaps in other situations there may be cases like this... anyway, that's not important now.
